File: python/tests/test_partition_predicates.py
#!/usr/bin/env python3
import fixtures
import partition_predicates as pps
import partition_tree as pt

import pytest
from hypothesis import given, settings, HealthCheck, assume
from hypothesis import strategies as st


@pytest.fixture
def p4D42():
    return fixtures.partitioners_list_4D_42()


@settings(suppress_health_check=[HealthCheck.function_scoped_fixture])
@given(idhbb=st.lists(st.integers(min_value=0, max_value=4),
                      min_size=4,
                      max_size=4),
       D=st.integers(min_value=1, max_value=4))
def test_only_bulk_and_borders(idhbb, D, p4D42):
    halo_dimensionality = sum(x in {0, 4} for x in idhbb)
    idx = [
        2,
        3,
        1,
        0,  #
        1,
        2,
        1,
        1,  #
        *idhbb,  #
        0,
        0
    ]  #

    res = halo_dimensionality <= D
    predicate = pps.get_NmD_halo_predicate(D)
    assert predicate(p4D42, idx) == res


@settings(suppress_health_check=[HealthCheck.function_scoped_fixture])
@given(
    MPI_rank_idx=st.lists(st.integers(min_value=0, max_value=3),
                          min_size=4,
                          max_size=4),
    MPI_rank_selected=st.lists(st.integers(min_value=0, max_value=3),
                               min_size=4,
                               max_size=4),
)
def test_MPI_selection_wrong(MPI_rank_idx, MPI_rank_selected, p4D42):
    assume(MPI_rank_idx != MPI_rank_selected)
    idx_wrong = [
        *MPI_rank_idx,  #
        1,
        2,
        1,
        1,  #
        1,
        2,
        0,
        3,  #
        0,
        0
    ]  #
    predicate = pps.get_mpi_rank_predicate(MPI_rank_selected)
    assert not predicate(p4D42, idx_wrong)


@settings(suppress_health_check=[HealthCheck.function_scoped_fixture])
@given(MPI_rank_idx=st.lists(st.integers(min_value=0, max_value=3),
                             min_size=4,
                             max_size=4))
def test_MPI_selection_right(MPI_rank_idx, p4D42):
    idx_right = [
        *MPI_rank_idx,  #
        1,
        2,
        1,
        1,  #
        1,
        2,
        0,
        3,  #
        0,
        0
    ]  #

    predicate = pps.get_mpi_rank_predicate(MPI_rank_idx)

    assert predicate(p4D42, idx_right)

def test_nobulk():
    partitioners = fixtures.partitioners_list_1D_42()
    partitioning = fixtures.partitioning1D()
    size_tree = pt.get_size_tree(
        partitioning, lambda idx: pps.no_bulk_borders(partitioners, idx))

    assert size_tree[0][0] == 16

File: cpp/tests/test_partition_predicates.cpp
#include "bool_maybe.hpp"
#include "fixtures.hpp"
#include "partition_predicates.hpp"
#include "partition_tree.hpp"
#include "test_utils.hpp"
#include <algorithm>
#include <boost/test/data/test_case.hpp>
#include <boost/test/unit_test.hpp>
using namespace hypercubes::slow;
using namespace hypercubes::slow::partitioners;
using namespace data;

BOOST_AUTO_TEST_SUITE(test_predicates)

BOOST_FIXTURE_TEST_CASE(test_only_bulk_and_borders, Part4DF) {
  auto data = rilist(0, 4, 4, 4);
  auto idhbb_it = data.begin();
  for (int _ = 0; _ < 200; ++_) {
    auto idhbb = *idhbb_it;
    ++idhbb_it;

    bool expected = std::all_of(idhbb.begin(), idhbb.end(),
                                [](int i) { return i != 0 and i != 4; });

    Indices idx{2,  //
                3,  //
                1,  //
                0,  //# End of MPI
                1,  //
                2,  //
                1,  //
                1}; //# End of Vector

    std::copy(idhbb.begin(), idhbb.end(), std::back_inserter(idx));
    idx.push_back(0);
    idx.push_back(0);

    BOOST_CHECK(to_bool(get_NmD_halo_predicate(partitioners, 0)(idx)) ==
                expected);
  }
}
BOOST_FIXTURE_TEST_CASE(test_mpi_selection_wrong, Part4DF) {
  auto data = rilist(0, 3, 4, 4);
  auto data_it = data.begin();
  for (int _ = 0; _ < 1000; ++_) {
    auto MPI_rank_idx = *data_it;
    ++data_it;
    auto MPI_rank_selected = *data_it;
    ++data_it;
    if (MPI_rank_idx == MPI_rank_selected)
      continue;

    Indices idx{-1, //
                -1, //
                -1, //
                -1, //# End of MPI
                1,  //
                2,  //
                1,  //
                1,  //# End of Vector
                1,  //
                2,  //
                0,  //
                3,  // # End of halos
                0,  //
                0};

    std::copy(MPI_rank_idx.begin(), //
              MPI_rank_idx.end(),   //
              idx.begin());

    auto predicate = get_mpi_rank_predicate(partitioners, MPI_rank_selected);
    BOOST_CHECK(not to_bool(predicate(idx)));
  }
}
BOOST_FIXTURE_TEST_CASE(test_mpi_selection_right, Part4DF) {
  auto data = rilist(0, 3, 4, 4);
  auto data_it = data.begin();
  for (int _ = 0; _ < 1000; ++_) {
    auto MPI_rank_idx = *data_it;
    ++data_it;

    Indices idx{-1, //
                -1, //
                -1, //
                -1, //# End of MPI
                1,  //
                2,  //
                1,  //
                1,  //# End of Vector
                1,  //
                2,  //
                0,  //
                3,  // # End of halos
                0,  //
                0};

    std::copy(MPI_rank_idx.begin(), //
              MPI_rank_idx.end(),   //
              idx.begin());

    auto predicate = get_mpi_rank_predicate(partitioners, MPI_rank_idx);
    BOOST_CHECK(to_bool(predicate(idx)));
  }
}

BOOST_AUTO_TEST_SUITE_END()
